lib/compilers/coccinelle.ts

```diff
@@ -0,0 +1,281 @@
+
+        if (!execOptions.customCwd) {
+            execOptions.customCwd = path.dirname(inputFilename);
+        }
+
+        const spatchFilename = path.join(path.dirname(inputFilename), this.spatchBaseFilename);
+        let cFileContents = '';
+        let pFileContents = '';
+        let toc = true; // toc: Target Language or Coccinelle Language
+        const file = await open(inputFilename);
+        for await (const line of file.readLines()) {
+            // this separates the C/C++ portion from the SmPL
+            const ifdefCocciRegex = /^#ifdef.*COCCINELLE.*/;
+            const endifCocciRegex = /^#endif.*COCCINELLE.*/;
+
+            if (ifdefCocciRegex.test(line) && toc) {
+                toc = !toc;
+            } else {
+                if (!endifCocciRegex.test(line)) {
+                    if (toc) cFileContents += line + '\n';
+                    else pFileContents += line + '\n';
+                }
+            }
+        }
+        file.close();
+
+        // we overwrite the C/C++ + SmPL with extracted C/C++ sources
+        fs.writeFileSync(inputFilename, cFileContents);
+        // we save the extracted SmPL sources
+        fs.writeFileSync(spatchFilename, pFileContents);
+
+        // check that the SmPL source does not #include anything, via CE rules
+        let cocciSourceError = this.checkSource(pFileContents);
+        if (cocciSourceError) throw cocciSourceError;
+        // use own rules, too
+        cocciSourceError = this.checkCocciSource(pFileContents);
+        if (cocciSourceError) throw cocciSourceError;
+
+        const result = await this.exec(compiler, options, execOptions);
+
+        if (this.joinSpatchStdinAndStderr && result.code === 0 && !result.timedOut) {
+            result.stderr += result.stdout;
+            result.stdout = result.stderr;
+            result.stderr = '';
+        }
+        return {
+            ...this.transformToCompilationResult(result, inputFilename),
+            languageId: this.getCompilerResultLanguageId(filters),
+            instructionSet: this.getInstructionSetFromCompilerArgs(options),
+        };
+    }
+
+    override getOutputFilename(dirPath: string, outputFilebase: string, key?: CacheKey): string {
+        return path.join(dirPath, `${outputFilebase}${this.lang.extensions[0]}`);
+    }
+
+    override postCompilationPreCacheHook(result: CompilationResult): CompilationResult {
+        if (result.code === 0) {
+            result.languageId = 'c'; // we produce C
+        }
+        return result;
+    }
+
+    override filterUserOptions(args: string[]) {
+        const permittedOptions = new Set(['--debug', '--verbose', '--quiet']); // preliminary and in need of a mention somewhere
+
+        return args.filter(item => {
+            if (typeof item !== 'string') return true;
+
+            if (permittedOptions.has(item)) return true;
+            else {
+                logger.warn(`User-provided option ${item} not allowed -- ignoring it.`);
```

Comment:
This could cause a large number of `warn`ings. On the production Compiler Explorer (if we were to host this) it would cause a lot of logging based on reasonable user input. Do you think this is a useful check and log line?

Reply:
No, there's no need to have this in production.
I added a conditional for this line, which by default is false (see https://github.com/compiler-explorer/compiler-explorer/pull/7001/commits/4e6778804708892ff5b9fca2761d18b1c069b85c)